ping/ping.md

```diff
@@ -0,0 +1,62 @@
+# Ping <!-- omit in toc -->
+
+| Lifecycle Stage | Maturity                 | Status | Latest Revision |
+| --------------- | ------------------------ | ------ | --------------- |
+| 1A              | Candidate Recommendation | Active | r0, 2022-11-04  |
+
+Authors: [@marcopolo]
+
+Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]
+
+[@marcopolo]: https://github.com/mxinden
+[@mxinden]: https://github.com/mxinden
+[@marten-seemann]: https://github.com/marten-seemann
+
+# Table of Contents <!-- omit in toc -->
+- [Protocol](#protocol)
+- [Diagram](#diagram)
+
+# Protocol
+
+The ping protocol is a simple request response protocol. The client opens a
+stream, sends a payload of 32 random bytes, and the server responds with the
+same 32 bytes on that stream. The client then measures the RTT from the time it
+wrote the bytes to the time it received the bytes. The client MAY repeat the
+process by sending another payload with random bytes on the same stream, so the
+server SHOULD loop and echo the next payload. The client SHOULD close the write
+side of the stream after sending the last payload, and the server SHOULD finish
+writing the echoed payload and then exit the loop and close the stream.
+
+The client MUST NOT keep more than one outbound stream for the ping protocol per
+peer. The server SHOULD accept at most 2 streams per peer since cross stream
+behavior is not linearizable for client and server. In other words, the client
+closing stream A and then opening stream B, might be perceived by the server as
+the client opening stream B and then closing stream A.
```

Comment:
So for instance

```
The server SHOULD wait 200 milliseconds after a ping before accepting another ping a stream.
The client MUST wait 400 milliseconds after a reply to ping the server again on a stream.
```

(or make it global instead of per stream, idk)

200ms of margin to allow for jitter at the networking & scheduler level. This is less elegant than just one request per stream, but what can we do. If your ping suddently goes from 500s to 100ms (for instance if the cpu is clogged up), you will have erroneous values for one ping.

Reply:
Why? This isn’t DOS mitigation. If you’re worried about rate limiting a 32byte response, then you probably want to do something at a higher level (rate limiting all streams and connections for example).